clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientCallbackHandler.java

```diff
@@ -78,10 +82,21 @@ public void handle(Callback[] callbacks) throws UnsupportedCallbackException {
                 if (ac.isAuthorized())
                     ac.setAuthorizedID(authzId);
             } else if (callback instanceof ScramExtensionsCallback) {
+                if (SaslConfigs.GSSAPI_MECHANISM.equals(mechanism))
+                    return; // extensions are not supported for GSSAPI
+
                 ScramExtensionsCallback sc = (ScramExtensionsCallback) callback;
-                if (!SaslConfigs.GSSAPI_MECHANISM.equals(mechanism) && subject != null && !subject.getPublicCredentials(Map.class).isEmpty()) {
+                if (subject != null && !subject.getPublicCredentials(Map.class).isEmpty()) {
                     sc.extensions((Map<String, String>) subject.getPublicCredentials(Map.class).iterator().next());
                 }
+            } else if (callback instanceof SaslExtensionsCallback) {
+                if (SaslConfigs.GSSAPI_MECHANISM.equals(mechanism))
+                    return; // extensions are not supported for GSSAPI
```

Comment:
Same as before - check the mechanism in the `if` statement below. Also, I think we could check for Scram mechanism in the check above and check for OAUTHBEARER here.